File: parcelabletest/src/main/java/com/backbase/test/instantiator/CompositeInstantiator.java
package com.backbase.test.instantiator;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import androidx.annotation.NonNull;

/**
 * Created by Backbase R&D B.V. on 10/12/2018.
 * An {@link Instantiator} that delegates to a list of {@link Instantiator}s.
 * If multiple delegates support the same type, the first one in the list is
 * used.
 */
public final class CompositeInstantiator extends TypeLimitedInstantiator {

    @NonNull private final List<Instantiator> delegates;

    public CompositeInstantiator(@NonNull List<Instantiator> delegates) {
        this.delegates = Collections.unmodifiableList(delegates);
    }

    public CompositeInstantiator(@NonNull Instantiator... delegates) {
        this(Arrays.asList(delegates));
    }

    @Override
    public boolean supports(@NonNull Class<?> objectClass) {
        for (Instantiator delegate : delegates) {
            if (delegate.supports(objectClass)) {
                return true;
            }
        }
        return false;
    }

    @Override
    protected <O> O instantiateSupportedType(@NonNull Class<O> objectClass) {
        for (Instantiator delegate : delegates) {
            if (delegate.supports(objectClass)) {
                return delegate.instantiate(objectClass);
            }
        }
        throw new IllegalStateException("Expected a supported type, but received " + objectClass.getName());
    }
}

File: parcelabletest/src/main/java/com/backbase/test/instantiator/CompositeInstantiator.kt
package com.backbase.test.instantiator

/**
 * Created by Backbase R&D B.V. on 2019-06-12.
 * A collection of multiple [Instantiator]s that can functions as a single [MultiTypeInstantiator]. If multiple [Instantiator]s provided during
 * construction support the same type, only the first one will be used.
 */
open class CompositeInstantiator(
    vararg instantiators: Instantiator<*>
) : MultiTypeInstantiator {

    constructor(instantiators: List<Instantiator<*>>) : this(*instantiators.toTypedArray())

    private val instantiatorsByType: Map<Class<*>, Instantiator<*>>

    init {
        val instantiatorsByType: MutableMap<Class<*>, Instantiator<*>> = mutableMapOf()
        for (instantiator in instantiators) {
            val type = instantiator.supportedClass()
            if (!instantiatorsByType.contains(type))
                instantiatorsByType[type] = instantiator
        }
        this.instantiatorsByType = instantiatorsByType
    }

    final override fun supports(type: Class<*>) = instantiatorsByType.contains(type)

    final override fun <T> instantiate(type: Class<T>): T {
        val instantiator = instantiatorsByType[type] ?: error("Type $type is not supported by this ${javaClass.simpleName}")
        @Suppress("UNCHECKED_CAST") return instantiator.instantiate() as T
    }
}
